File: Samples/Evaluate/how-to-implement-the-basic-idea-of-the-microsoft-excel-format-painter-feature-for-spreadsh-t590741-19.2.2-/CS/WindowsFormsApplication1/FormatPainterProvider.cs
using DevExpress.Spreadsheet;
using DevExpress.XtraBars;
using DevExpress.XtraSpreadsheet;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Windows.Forms;

namespace WindowsFormsApplication1
{
    public class FormatPainterProvider
    {
        SpreadsheetControl spreadsheetControl;
        BarCheckItem biFormatPainter;
        CellRange sourceCell = null;
        FormatPainterMode formatPainterMode = FormatPainterMode.None;
        enum FormatPainterMode { SingleAction, MultipleActions, None }
        public void RegisterFormatPainter(SpreadsheetControl spreadsheet, BarCheckItem biFormatPainter)
        {
            spreadsheetControl = spreadsheet;
            this.biFormatPainter = biFormatPainter;
            spreadsheet.MouseUp += spreadsheetControl1_MouseUp;
            spreadsheet.PreviewKeyDown += spreadsheetControl1_PreviewKeyDown;
            spreadsheet.SelectionChanged += spreadsheetControl1_SelectionChanged;
            spreadsheet.CellBeginEdit += spreadsheetControl1_CellBeginEdit;
            spreadsheet.SheetRemoving += Spreadsheet_SheetRemoving;
            biFormatPainter.CheckedChanged += barCheckItem1_CheckedChanged;
            biFormatPainter.ItemDoubleClick += barCheckItem1_ItemDoubleClick;

        }
        private void Spreadsheet_SheetRemoving(object sender, SheetRemovingEventArgs e)
        {
            if (sourceCell != null && e.SheetName == sourceCell.Worksheet.Name)
                biFormatPainter.Checked = false;
        }
        private void barCheckItem1_CheckedChanged(object sender, DevExpress.XtraBars.ItemClickEventArgs e)
        {
            bool shouldCopyFormat = biFormatPainter.Checked;
            sourceCell = shouldCopyFormat ? spreadsheetControl.SelectedCell : null;

            formatPainterMode = shouldCopyFormat ? FormatPainterMode.SingleAction : FormatPainterMode.None;
        }

        private void spreadsheetControl1_MouseUp(object sender, MouseEventArgs e)
        {
            if (IsFormatPainterActivated())
                ApplyFormat();
        }

        private bool IsFormatPainterActivated()
        {
            if (formatPainterMode == FormatPainterMode.None || sourceCell == null)
                return false;

            return true;
        }
        private void ApplyFormat()
        {
            spreadsheetControl.Selection.CopyFrom(sourceCell, PasteSpecial.Formats);
            if (formatPainterMode == FormatPainterMode.SingleAction)
                biFormatPainter.Checked = false;
        }

        private void barCheckItem1_ItemDoubleClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e)
        {
            biFormatPainter.Checked = true;
            formatPainterMode = FormatPainterMode.MultipleActions;
        }

        private void spreadsheetControl1_PreviewKeyDown(object sender, PreviewKeyDownEventArgs e)
        {
            if (!IsFormatPainterActivated()) return;

            if (e.KeyCode == Keys.Escape)
            {
                biFormatPainter.Checked = false;
            }
        }

        private void spreadsheetControl1_SelectionChanged(object sender, EventArgs e)
        {
            if (spreadsheetControl.SelectedShape != null || spreadsheetControl.SelectedComment != null)
            {
                biFormatPainter.Enabled = false;
                biFormatPainter.Checked = false;
            }
            else
                biFormatPainter.Enabled = true;
        }

        private void spreadsheetControl1_CellBeginEdit(object sender, DevExpress.XtraSpreadsheet.SpreadsheetCellCancelEventArgs e)
        {
            if (IsFormatPainterActivated())
            {
                ApplyFormat();
                e.Cancel = true;
            }
            else
                biFormatPainter.Enabled = false;
        }
        
    }
    
}

File: Samples/Evaluate/how-to-implement-the-basic-idea-of-the-microsoft-excel-format-painter-feature-for-spreadsh-t590741-19.2.2-/VB/WindowsFormsApplication1/FormatPainterProvider.vb
Imports DevExpress.Spreadsheet
Imports DevExpress.XtraBars
Imports DevExpress.XtraSpreadsheet
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms

Namespace WindowsFormsApplication1
	Public Class FormatPainterProvider
		Private spreadsheetControl As SpreadsheetControl
		Private biFormatPainter As BarCheckItem
		Private sourceCell As CellRange = Nothing
'INSTANT VB NOTE: The field formatPainterMode was renamed since Visual Basic does not allow fields to have the same name as other class members:
		Private formatPainterMode_Renamed As FormatPainterMode = WindowsFormsApplication1.FormatPainterProvider.FormatPainterMode.None
		Private Enum FormatPainterMode
			SingleAction
			MultipleActions
			None
		End Enum
		Public Sub RegisterFormatPainter(ByVal spreadsheet As SpreadsheetControl, ByVal biFormatPainter As BarCheckItem)
			spreadsheetControl = spreadsheet
			Me.biFormatPainter = biFormatPainter
			AddHandler spreadsheet.MouseUp, AddressOf spreadsheetControl1_MouseUp
			AddHandler spreadsheet.PreviewKeyDown, AddressOf spreadsheetControl1_PreviewKeyDown
			AddHandler spreadsheet.SelectionChanged, AddressOf spreadsheetControl1_SelectionChanged
			AddHandler spreadsheet.CellBeginEdit, AddressOf spreadsheetControl1_CellBeginEdit
			AddHandler spreadsheet.SheetRemoving, AddressOf Spreadsheet_SheetRemoving
			AddHandler biFormatPainter.CheckedChanged, AddressOf barCheckItem1_CheckedChanged
			AddHandler biFormatPainter.ItemDoubleClick, AddressOf barCheckItem1_ItemDoubleClick

		End Sub
		Private Sub Spreadsheet_SheetRemoving(ByVal sender As Object, ByVal e As SheetRemovingEventArgs)
			If sourceCell IsNot Nothing AndAlso e.SheetName = sourceCell.Worksheet.Name Then
				biFormatPainter.Checked = False
			End If
		End Sub
		Private Sub barCheckItem1_CheckedChanged(ByVal sender As Object, ByVal e As DevExpress.XtraBars.ItemClickEventArgs)
			Dim shouldCopyFormat As Boolean = biFormatPainter.Checked
			sourceCell = If(shouldCopyFormat, spreadsheetControl.SelectedCell, Nothing)

			formatPainterMode_Renamed = If(shouldCopyFormat, WindowsFormsApplication1.FormatPainterProvider.FormatPainterMode.SingleAction, WindowsFormsApplication1.FormatPainterProvider.FormatPainterMode.None)
		End Sub

		Private Sub spreadsheetControl1_MouseUp(ByVal sender As Object, ByVal e As MouseEventArgs)
			If IsFormatPainterActivated() Then
				ApplyFormat()
			End If
		End Sub

		Private Function IsFormatPainterActivated() As Boolean
			If formatPainterMode_Renamed = WindowsFormsApplication1.FormatPainterProvider.FormatPainterMode.None OrElse sourceCell Is Nothing Then
				Return False
			End If

			Return True
		End Function
		Private Sub ApplyFormat()
			spreadsheetControl.Selection.CopyFrom(sourceCell, PasteSpecial.Formats)
			If formatPainterMode_Renamed = WindowsFormsApplication1.FormatPainterProvider.FormatPainterMode.SingleAction Then
				biFormatPainter.Checked = False
			End If
		End Sub

		Private Sub barCheckItem1_ItemDoubleClick(ByVal sender As Object, ByVal e As DevExpress.XtraBars.ItemClickEventArgs)
			biFormatPainter.Checked = True
			formatPainterMode_Renamed = WindowsFormsApplication1.FormatPainterProvider.FormatPainterMode.MultipleActions
		End Sub

		Private Sub spreadsheetControl1_PreviewKeyDown(ByVal sender As Object, ByVal e As PreviewKeyDownEventArgs)
			If Not IsFormatPainterActivated() Then
				Return
			End If

			If e.KeyCode = Keys.Escape Then
				biFormatPainter.Checked = False
			End If
		End Sub

		Private Sub spreadsheetControl1_SelectionChanged(ByVal sender As Object, ByVal e As EventArgs)
			If spreadsheetControl.SelectedShape IsNot Nothing OrElse spreadsheetControl.SelectedComment IsNot Nothing Then
				biFormatPainter.Enabled = False
				biFormatPainter.Checked = False
			Else
				biFormatPainter.Enabled = True
			End If
		End Sub

		Private Sub spreadsheetControl1_CellBeginEdit(ByVal sender As Object, ByVal e As DevExpress.XtraSpreadsheet.SpreadsheetCellCancelEventArgs)
			If IsFormatPainterActivated() Then
				ApplyFormat()
				e.Cancel = True
			Else
				biFormatPainter.Enabled = False
			End If
		End Sub

	End Class

End Namespace
